dhctl/pkg/kubernetes/actions/converge/nodebootstrap.go

```diff
@@ -72,6 +77,121 @@ func BootstrapAdditionalNode(kubeCl *client.KubernetesClient, cfg *config.MetaCo
 	return nil
 }
 
+func BootstrapAdditionalNodeForParallelRun(kubeCl *client.KubernetesClient, cfg *config.MetaConfig, index int, step, nodeGroupName, cloudConfig string, isConverge bool, terraformContext *terraform.TerraformContext, buff *bytes.Buffer, saveLogToBuffer bool) error {
+	nodeName := NodeName(cfg, nodeGroupName, index)
+
+	nodeGroupSettings := cfg.FindTerraNodeGroup(nodeGroupName)
+
+	// TODO pass cache as argument or better refact func
+	runner := terraformContext.GetBootstrapNodeRunner(cfg, cache.Global(), terraform.BootstrapNodeRunnerOptions{
+		AutoApprove:     true,
+		NodeName:        nodeName,
+		NodeGroupStep:   step,
+		NodeGroupName:   nodeGroupName,
+		NodeIndex:       index,
+		NodeCloudConfig: cloudConfig,
+		LogToBuffer:     saveLogToBuffer,
+
+		AdditionalStateSaverDestinations: []terraform.SaverDestination{
+			NewNodeStateSaver(kubeCl, nodeName, nodeGroupName, nodeGroupSettings),
+		},
+	})
+
+	outputs, err := terraform.ApplyPipeline(runner, nodeName, terraform.OnlyState)
+	if err != nil {
+		return err
+	}
+
+	if tomb.IsInterrupted() {
+		return ErrConvergeInterrupted
+	}
+
+	err = SaveNodeTerraformState(kubeCl, nodeName, nodeGroupName, outputs.TerraformState, nodeGroupSettings)
+	if err != nil {
+		return err
+	}
+
+	if saveLogToBuffer {
+		logs := runner.GetLog()
+		buff.WriteString(strings.Join(logs, "\n"))
+	}
+
+	return nil
+}
+
+func ParallelBootstrapAdditionalNodes(kubeCl *client.KubernetesClient, cfg *config.MetaConfig, nodesIndexToCreate []int, step, nodeGroupName, cloudConfig string, isConverge bool, terraformContext *terraform.TerraformContext) ([]string, error) {
+
+	var (
+		nodesToWait []string
+		wg          sync.WaitGroup
+	)
+
+	type checkResult struct {
+		name    string
+		buffLog *bytes.Buffer
+		err     error
+	}
+
+	for _, indexCandidate := range nodesIndexToCreate {
+		candidateName := fmt.Sprintf("%s-%s-%v", cfg.ClusterPrefix, nodeGroupName, indexCandidate)
+		nodeExists, err := IsNodeExistsInCluster(kubeCl, candidateName)
+		if err != nil {
+			return nil, err
+		} else if nodeExists {
+			return nil, fmt.Errorf("node with name %s exists in cluster", candidateName)
+		}
+	}
+
+	if len(nodesIndexToCreate) > 1 {
+		log.WarnF("Many pipelines will run in parallel, terraform output for nodes %s-%v will be displayed after main execution.\n\n", nodeGroupName, nodesIndexToCreate[1:])
+	}
+
+	resultsChan := make(chan checkResult, len(nodesIndexToCreate))
+	for i, indexCandidate := range nodesIndexToCreate {
+		saveLogToBuffer := true
+		candidateName := fmt.Sprintf("%s-%s-%v", cfg.ClusterPrefix, nodeGroupName, indexCandidate)
+		var buffLog bytes.Buffer
+		wg.Add(1)
+		go func() {
```

Comment:
better way pass index as argument and use it than catch it with closure

Reply:
done